refactor(src): change verbose messages

TDA opened the pull request:
move reporting logic close to bloom filter logic
created a variable called URl to hold the directory data
added a check for password to be string and non-empty
removed xhr calls
added grunt and requirejs
created build package with minified file
refactored function and variable names
added new error messages
added tests for error messages
Fixes #1 #2 #3 #4 #5 #6 #7 #8 #9 #10 includes #11 as part of this.
Well, that is a mouthful.
Can be merged :D :D Yay!

TDA committed:
refactor(src): change variable and function name to better reflect meaning
add bower to package.json
escape from eslints clutches
edit travis.yml

TDA committed:
refactor(src): change verbose messages
move reporting logic close to bloom filter logic
created a variable called URl to hold the directory data
added check for password to be a string and non-empty
refactored some variable names and the conditions
added unit tests for the new messages
changed unit tests for bloom filter

shane-tomlinson commented on src/passwordcheck.js:
```diff
@@ -28,29 +36,26 @@ function ($, bloomfilter) {
-      callback('Password with all alphabets/numbers detected');
-    }
+  return function (password, callback) {
+    if (password && typeof password === 'string' && password.length > 8) {
```
Maybe the check should take place on all string passwords, and if `password.length < 8`, then callback with a new status code `TOO_SHORT`?

shane-tomlinson replied:
Addendum, maybe we could use a few more error codes here for completeness.

If there is no password, `PASSWORD_NOT_GIVEN`
If the password is not a string, `PASSWORD_NOT_A_STRING`
If the password is too short: `PASSWORD_TOO_SHORT`

TDA replied:
@shane-tomlinson The user wont be able to submit the form unless those conditions are met, do we still need the status codes? I can add them in easily.

shane-tomlinson replied:
I am thinking towards the future when we can release this library to the world! Other folks won't have the same form submission logic as us.

shane-tomlinson commented on src/passwordcheck.js:
```diff
@@ -28,29 +36,26 @@ function ($, bloomfilter) {
-    }
+  return function (password, callback) {
+    if (password && typeof password === 'string' && password.length > 8) {
+      var isPasswordWeak = false;
```
`isPasswordWeak` doesn't really serve a purpose way out here, nor on line 56 since nothing is done with the value after it is set. It's only really useful on lines 49 & 50. I'd move the `var` declaration to line 49.

TDA replied:
done!

shane-tomlinson commented on src/passwordcheck.js:
```diff
@@ -28,29 +36,26 @@ function ($, bloomfilter) {
+  return function (password, callback) {
+    if (password && typeof password === 'string' && password.length > 8) {
+      var isPasswordWeak = false;
+      var strong = strengthCheck(password);
```
Can you rename `strong` to `isStrong`? I like to preface boolean values with something that indicates it's a truthy value, `is`, `has`, `can`, etc.

TDA replied:
done!

TDA commented:
one ultra customizable password checker ready @shane-tomlinson  :D :D

shane-tomlinson commented on src/passwordcheck.js:
```diff
@@ -10,47 +10,71 @@ define(['jquery',
+    return function (password, callback) {
+      if (! password) {
+        callback(MESSAGES.MISSING_PASSWORD);
+      } else if (typeof password !== 'string'){
```
nit: space between `)` and `{`

shane-tomlinson replied:
Note, I'm surprised `npm run eslint` didn't catch this. I'm also surprised if I run `eslint .` from the root directory, it blows up.

shane-tomlinson commented on src/passwordcheck.js:
```diff
@@ -10,47 +10,71 @@ define(['jquery',
+    function strengthCheck(password) {
+      // Check for passwords that at least contain a number and an alphabet,
+      // or if alphabets, then at least 12 characters long
+      var regex = new RegExp('((?=.*[a-zA-Z])(?=.*[0-9])[A-Za-z0-9!@#$%^&*()_+ ]{8,12})|([A-Za-z0-9!@#$%^&*()_+ ]{12,})');
```
Since `minLength` is now configurable, does the `8` in `{8,12}` need to be updated to use `minLength`?

shane-tomlinson commented on src/passwordcheck.js:
```diff
@@ -10,47 +10,71 @@ define(['jquery',
+
+    return function (password, callback) {
+      if (! password) {
+        callback(MESSAGES.MISSING_PASSWORD);
```
Unit tests are needed for all these new cases.

shane-tomlinson commented on src/passwordcheck.js:
```diff
@@ -10,47 +10,71 @@ define(['jquery',
-      callback('Password found in bloom filter');
-    } else if (!bool && strong) {
-      callback('Password was not found in bloom filter');
+    $.ajax({
```
For a separate PR, since the content server is requiring the library asynchronously, making yet another XHR request to load the data is a bit overboard. I think a better approach would be to include grunt with this library and create a single bundle that can then be used in the content server and loaded asynchronously. Then we wouldn't have to worry about the bloom filter being available before using it, we could get rid of this XHR stuff, and passing in the alternative URL.

shane-tomlinson replied:
Note, creating a single bundle out of the source will require setting up grunt with requirejs. The fxa-js-client does it, [use their work as a guide, place to start](https://github.com/mozilla/fxa-js-client/blob/master/tasks/require.js).

TDA replied:
@shane-tomlinson Wouldnt loading a JSON file anyway require an ajax call? I dont understand how creating a single bundle would solve that for us :( Do you mean we create a single file with both the data and the script?

shane-tomlinson replied:
> @shane-tomlinson Wouldnt loading a JSON file anyway require an ajax call? I dont understand how creating a single bundle would solve that for us :( Do you mean we create a single file with both the data and the script?

Yes, we already load the library from the content server asynchronously, calling $.ajax here would slow things down by doing a 2nd XHR request.

shane-tomlinson replied:
So, what I am saying - bundle them together.

shane-tomlinson commented on package.json:
```diff
@@ -8,7 +8,7 @@
   },
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1",
+    "test": "echo \"Error: no test specified\" && exit 0",
```
At a minimum run the eslint script.

TDA commented:
@shane-tomlinson Added tests for new messages, removed the xhr call, and modified the regex to take minLength into account. Also tried a requireconfig here, if it works, i can move this to a require-config.js file here, and we can remove the config stuff in the content-server.

TDA committed:
chore(grunt): add grunt to repo
run build tasks

pdehaan commented on Gruntfile.js:
```diff
@@ -0,0 +1,25 @@
+  // this one is def not ready yet
+  grunt.registerTask('build',
+    'Build client',
+    ['clean', 'lint', 'requirejs']);
```
Do we have a 'clean' task?

TDA replied:
I did say it wasnt ready yet in the comment above that. @pdehaan

TDA committed:
chore(build): create build package for repo

TDA commented:
@shane-tomlinson Is that what you meant by a single package? I had the requirejs thing create a single file with all the amd modules in it. I mostly used the methods used in fxa-js-client. So now, we have no extraneous calls to other files. Self-contained.

TDA commented:
@shane-tomlinson One thing to be kept in mind while packaging them as one bundle is that we wont be able to change the no_of_hashing fns, should I remove that option?

TDA committed:
chore(build): finalized build package

shane-tomlinson commented on Gruntfile.js:
```diff
@@ -0,0 +1,27 @@
+  require('load-grunt-tasks')(grunt);
+  // load local Grunt tasks
+  grunt.loadTasks('tasks');
+  // this one is def not ready yet
```
is this a TODO?

shane-tomlinson commented on Gruntfile.js:
```diff
@@ -0,0 +1,27 @@
+  // load local Grunt tasks
+  grunt.loadTasks('tasks');
+  // this one is def not ready yet
+  grunt.registerTask('build',
```
Since you are loading the tasks from the `tasks` subdirectory, you might as well create two new files for the `build` and `lint` tasks to keep things consistent.

TDA replied:
will do :) and yes, ill remove the comment.

shane-tomlinson commented on tasks/eslint.js:
```diff
@@ -0,0 +1,19 @@
+      eslintrc: '.eslintrc',
+      src: [
+        'tests/*.js'
+      ]
```
I see the following warning when running `grunt lint`:

``` bash
➜  fxa-password-strength-checker git:(issue-3-verbose) grunt lint
Running "eslint:app" (eslint) task

src/bloomdata_short_pwd.js
  0:0  warning  File ignored because of your .eslintignore file. Use --no-ignore to override

✖ 1 problem (0 errors, 1 warning)
```

We generally treat warnings like errors - they shouldn't be there.

TDA replied:
I will get that fixed :)

shane-tomlinson commented on tests/index.html:
```diff
@@ -9,26 +9,35 @@
 <div id="mocha"></div>
 <script src="../bower_components/chai/chai.js"></script>
```
Instead of including these files directly and declaring globals, you should allow requirejs to include them. Whenever they are needed by a test module, include the dependency in the list of dependencies.

You'll probably have to declare the globals files like `chai` and `mocha` create. For an example, see [how the content server does it](https://github.com/mozilla/fxa-content-server/blob/c39d365ef4eb734777d947d6bf4e20b578a096cb/app/scripts/require_config.js#L45).

shane-tomlinson commented:
I'm going to merge this because this work is a huge step forward. I'll create separate issues for the other suggestions.

pdehaan commented:
Should the `build/*` folder be checked in to this repo?
